### PR TITLE
stfio build guide for non-default Python distributions on Mac and Linux

### PR DESCRIPTION
**Pull request by yueqiw**

### CODE REVIEW

File: doc/sphinx/linux_install_guide/moduleonly.rst [hunk 1]
```diff
@@ -41,6 +41,12 @@ Now you can configure. I strongly recommend building in a separate directory.
 Remember to add the argument *--with-biosig* to the configure script 
 if you want to have extra biomedical fileformats for stfio. 
 
+If you want to install the *stfio* module for a non-default Python distribution, such as **Anaconda Python**, use the argument *--prefix=* to specify the installation path as the *site-packages* folder of your favorite Python distribution. The following example will install *stfio* for Anaconda Python 2.7:
+
+::
+
+    $ ../../configure --enable-module --prefix=$HOME/anaconda/lib/python2.7/site-packages
```
Comment (neurodroid) [inline doc/sphinx/linux_install_guide/moduleonly.rst]: Don't you also need to specify `$ PYTHON=/where/your/anaconda/python/lives ../../configure --enable-module --prefix=$HOME/anaconda/lib/python2.7/site-packages`, or are you assuming that it's the first `python` in your path?

**Comment by neurodroid** [comment]
Have you tried to use the `setup.py` script with anaconda python?

File: doc/sphinx/linux_install_guide/moduleonly.rst (hunk 1 above)
Comment (yueqiw) [inline doc/sphinx/linux_install_guide/moduleonly.rst]: It worked for me without having to specify `PYTHON=`. In my case the default Python is already Anaconda: `which python` points to the `$HOME/anaconda/bin/python`. Do you think `PYTHON=` would be required for other cases?

**Comment by yueqiw** [comment]
@neurodroid I tried setup.py at the beginning but it did not work for me.

Here is the error message: 
```
ld: library not found for -lmkl_lapack95_lp64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
error: Command "g++ -bundle -undefined dynamic_lookup -L/Users/yueqi/anaconda/lib -arch x86_64 -arch x86_64 build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/abflib.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAbfFio32/Oldheadr.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAbfFio32/abferror.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAbfFio32/abffiles.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAbfFio32/abfheadr.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAbfFio32/abfhwave.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAbfFio32/abfutil.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAbfFio32/csynch.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAbfFio32/filedesc.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAbfFio32/msbincvt.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAtfFio32/axatffio32.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/AxAtfFio32/fileio2.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/Common/FileIO.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/Common/FileReadCache.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon/Common/unix.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon2/ProtocolReaderABF2.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon2/SimpleStringCache.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/abf/axon2/abf2headr.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/atf/atflib.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/axg/AxoGraph_ReadWrite.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/axg/axglib.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/axg/byteswap.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/axg/fileUtils.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/axg/stringUtils.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/biosig/biosiglib.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/cfs/cfs.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/cfs/cfslib.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/channel.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/hdf5/hdf5lib.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/igor/CrossPlatformFileIO.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/igor/WriteWave.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/igor/igorlib.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/recording.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/section.o build/temp.macosx-10.6-x86_64-2.7/src/libstfio/stfio.o build/temp.macosx-10.6-x86_64-2.7/src/libstfnum/fit.o build/temp.macosx-10.6-x86_64-2.7/src/libstfnum/funclib.o build/temp.macosx-10.6-x86_64-2.7/src/libstfnum/levmar/Axb.o build/temp.macosx-10.6-x86_64-2.7/src/libstfnum/levmar/lm.o build/temp.macosx-10.6-x86_64-2.7/src/libstfnum/levmar/lmbc.o build/temp.macosx-10.6-x86_64-2.7/src/libstfnum/levmar/misc.o build/temp.macosx-10.6-x86_64-2.7/src/libstfnum/measure.o build/temp.macosx-10.6-x86_64-2.7/src/libstfnum/stfnum.o build/temp.macosx-10.6-x86_64-2.7/src/pystfio/pystfio.o build/temp.macosx-10.6-x86_64-2.7/src/pystfio/pystfio_wrap.o -L/Users/yueqi/anaconda/lib -lhdf5 -lhdf5_hl -lfftw3 -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -liomp5 -lpthread -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -liomp5 -lpthread -lmkl_lapack95_lp64 -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -liomp5 -lpthread -lmkl_lapack95_lp64 -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -liomp5 -lpthread -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -liomp5 -lpthread -lbiosig2 -o build/lib.macosx-10.6-x86_64-2.7/_stfio.so" failed with exit status 1
```

From [this thread](https://github.com/conda/conda/issues/2308) it seems that `install nomkl` might fix this error (not sure if there will be other errors after this). But doing so would change numpy/scipy to nomkl versions.

**Comment by neurodroid** [comment]
Is there any file called `libmkl_lapack95_lp64.dylib` or similar in your anaconda directory?

**Comment by yueqiw** [comment]
`Python setup.py install` worked after installing nomkl package. 

There is no `libmkl_lapack95_lp64.dylib`, but I have `libmkl_intel_lp64.dylib` in the conda directory.

**Comment by neurodroid** [comment]
Thanks for looking into this! It would be nice to get this working with the mkl version though.
I don't think we need to link against `libmkl_lapack95_lp64.dylib` at all, even if we use the mkl version.

Could you try to comment out the lines in setup.py that add some numpy-specific libraries (lines 35 and 48) and try again with the mkl version?

Thanks for looking into this.

**Comment by yueqiw** [comment]
Now the problem is fixed! No need to change `setup.py`.

Actually the mkl issue is a problem with my mkl installation. Here is the output of numpy.show_config() when I run into `ld: library not found for -lmkl_lapack95_lp64` :
```
>>> import numpy as np
>>> np.show_config()
lapack_opt_info:
    libraries = ['mkl_lapack95_lp64', 'mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/include']
blas_opt_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/include']
openblas_lapack_info:
  NOT AVAILABLE
lapack_mkl_info:
    libraries = ['mkl_lapack95_lp64', 'mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/include']
blas_mkl_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/include']
mkl_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/include']
```
You can see the `mkl_lapack95_lp64` entry, although it's not present in the system. 
After `conda update mkl`, the problem is solved. I can install `stfio` by `python setup.py install` without issues. And here is the updated mkl config:
```
>>> np.show_config()
lapack_opt_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/include']
blas_opt_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/include']
openblas_lapack_info:
  NOT AVAILABLE
lapack_mkl_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/include']
blas_mkl_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/yueqi/anaconda/envs/stfmkl/include']
```

However, the first time I `import stfio` after install, I got the following error:
```
Python 2.7.12 |Anaconda 4.2.0 (x86_64)| (default, Jul  2 2016, 17:43:17) 
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import stfio
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/stfio/__init__.py", line 6, in <module>
    from .stfio import *
  File "/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/stfio/stfio.py", line 23, in <module>
    _stfio = swig_import_helper()
  File "/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/stfio/stfio.py", line 22, in swig_import_helper
    return importlib.import_module('_stfio')
  File "/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
ImportError: dlopen(/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/_stfio.so, 2): Symbol not found: _cholmod_finish
  Referenced from: /Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/_stfio.so
  Expected in: flat namespace
 in /Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/_stfio.so
```

But this does not seem to be a problem. All you need to do is to restart the terminal. 
After I restart the terminal, the error is gone. 
```
>>> import stfio
>>> 
```

**Comment by yueqiw** [comment]
Actually the `ImportError` after installation persists. Sometimes I can load the stfio library, sometimes I still get the same error. 

```
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-0ed2f487f8cd> in <module>()
      1 import numpy as np
      2 import matplotlib.pyplot as plt
----> 3 import stfio
      4 import os

/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/stfio/__init__.py in <module>()
      4 '''
      5 
----> 6 from .stfio import *
      7 from . import stfio_plot as plot
      8 try:

/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/stfio/stfio.py in <module>()
     21         except ImportError:
     22             return importlib.import_module('_stfio')
---> 23     _stfio = swig_import_helper()
     24     del swig_import_helper
     25 elif _swig_python_version_info >= (2, 6, 0):

/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/stfio/stfio.py in swig_import_helper()
     20             return importlib.import_module(mname)
     21         except ImportError:
---> 22             return importlib.import_module('_stfio')
     23     _stfio = swig_import_helper()
     24     del swig_import_helper

/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/importlib/__init__.pyc in import_module(name, package)
     35             level += 1
     36         name = _resolve_name(name[level:], package, level)
---> 37     __import__(name)
     38     return sys.modules[name]

ImportError: dlopen(/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/_stfio.so, 2): Symbol not found: _cholmod_finish
  Referenced from: /Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/_stfio.so
  Expected in: flat namespace
 in /Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/_stfio.so
```

**Comment by neurodroid** [comment]
@yueqiw Can you please pull the latest version of the setup.py script and try again (make sure you've deleted your build directory). I've gotten rid of these dependencies.

**Comment by yueqiw** [comment]
The install works out without errors, but I still got a similar error when I import the module. 

```
Python 2.7.12 |Anaconda 4.2.0 (x86_64)| (default, Jul  2 2016, 17:43:17) 
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import stfio
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/stfio/__init__.py", line 6, in <module>
    from .stfio import *
  File "/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/stfio/stfio.py", line 23, in <module>
    _stfio = swig_import_helper()
  File "/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/stfio/stfio.py", line 22, in swig_import_helper
    return importlib.import_module('_stfio')
  File "/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
ImportError: dlopen(/Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/_stfio.so, 2): Symbol not found: _libiconv
  Referenced from: /Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/_stfio.so
  Expected in: flat namespace
 in /Users/yueqi/anaconda/envs/stfmkl/lib/python2.7/site-packages/_stfio.so
>>> 
```

**Pull request merged**